### PR TITLE
RR-320 - Support navigating to Create A Goal directly from CIAG

### PR DESCRIPTION
**Pull request by nathanrussell-moj-digital**
This PR changes our Create A Goal process so that it can be navigated to directly from CIAG screens.

Previously there was a request handler middleware that ensured the user had come from our (PLP) Overview screen. This process no longer works / is appropriate now that the CIAG screens need to link directly to our Create A Goal process

### CODE REVIEW

File: integration_tests/e2e/goal/createGoal.cy.ts [hunk 1]
```diff
@@ -15,7 +15,22 @@ context('Create a goal', () => {
     cy.task('stubLearnerEducation')
   })
 
-  it('should not be able to navigate directly to Create Goal page given user has not clicked Add A Goal from overview page', () => {
+  it('should be able to navigate to Create Goal page given user has clicked Add A Goal from Overview page', () => {
```
Comment (nathanrussell-moj-digital) [inline integration_tests/e2e/goal/createGoal.cy.ts:18]: Confirm that existing behaviour of clicking Create Goal on PLP's overview screen still works

File: integration_tests/e2e/goal/createGoal.cy.ts [hunk 2]
```diff
@@ -15,7 +15,22 @@ context('Create a goal', () => {
...
+    const overviewPage = Page.verifyOnPage(OverviewPage)
+
+    // When
+    const createGoalPage = overviewPage.clickAddGoalButton()
+
+    // Then
+    createGoalPage.isForPrisoner(prisonNumber)
+  })
+
+  it('should be able to navigate directly to Create Goal page from non-PLP pages such as CIAG service', () => {
```
Comment (nathanrussell-moj-digital) [inline integration_tests/e2e/goal/createGoal.cy.ts:33]: Test to prove that we can now navigate directly to Create Goal without going via PLP's Overview screen

File: server/routes/createGoal/index.ts [hunk 3]
```diff
@@ -17,7 +18,7 @@ export default (router: Router, services: Services) => {
 
   router.use('/plan/:prisonNumber/goals/create', checkUserHasEditAuthority())
   router.get('/plan/:prisonNumber/goals/create', [
-    checkPrisonerSummaryExistsInSession,
```
Comment (nathanrussell-moj-digital) [inline server/routes/createGoal/index.ts:20]: This is the request handler that we used previously for the GET route for creating a goal. It checks whether the prisoner summary is in session, and if no redirects the user to PLP's Overview page

The request handler `checkPrisonerSummaryExistsInSession` is still in use in most of our other routes, its just that this route now needs to do something slightly different; hence the change to use the new request handler 👉

File: server/routes/functionalSkills/index.ts [hunk 4]
```diff
@@ -1,20 +1,18 @@
...
  * Route definitions for the pages relating to Functional Skills
  */
 export default (router: Router, services: Services) => {
-  const prisonerSummaryRequestHandler = new PrisonerSummaryRequestHandler(services.prisonerSearchService)
   const functionalSkillsController = new FunctionalSkillsController(services.curiousService)
 
   router.use('/plan/:prisonNumber/functional-skills', [
     checkUserHasViewAuthority(),
-    prisonerSummaryRequestHandler.getPrisonerSummary,
+    retrievePrisonerSummaryIfNotInSession(services.prisonerSearchService),
```
Comment (nathanrussell-moj-digital) [inline server/routes/functionalSkills/index.ts:15]: The new `retrievePrisonerSummaryIfNotInSession` request handler is basically a re-work / improvement of the existing `prisonerSummaryRequestHandler.getPrisonerSummary`; hence the change here to use the new request handler
(no point having 2 request handlers that do the same thing)

File: server/routes/routerRequestHandlers.test.ts [hunk 5]
```diff
@@ -371,4 +377,144 @@ describe('routerRequestHandlers', () => {
       expect(next).not.toHaveBeenCalled()
     })
   })
+
+  describe('retrievePrisonerSummaryIfNotInSession', () => {
```
Comment (nathanrussell-moj-digital) [inline server/routes/routerRequestHandlers.test.ts:381]: All of these tests were moved from the now deleted `server/routes/overview/prisonerSummaryRequestHandler.test.ts`

File: server/routes/routerRequestHandlers.ts [hunk 6]
```diff
@@ -105,10 +108,39 @@ const checkNewGoalsFormExistsInSession = async (req: Request, res: Response, nex
   }
 }
 
+/**
+ *  Middleware function that returns a Request handler function to look up the prisoner from prisoner-search, map to a PrisonerSummary, and store in the session
+ */
+const retrievePrisonerSummaryIfNotInSession = (prisonerSearchService: PrisonerSearchService): RequestHandler => {
```
Comment (nathanrussell-moj-digital) [inline server/routes/routerRequestHandlers.ts:114]: This new middleware function is a rework/improvement of what was previously in the now deleted `server/routes/overview/prisonerSummaryRequestHandler.ts`

**Review by mattwills-moj-digital — APPROVED** [review]

**Pull request merged**